crates/mdb/src/lib.rs

```diff
@@ -21,87 +20,37 @@ macro_rules! suppress_unwind {
     };
 }
 
-type OnMessage = dyn FnMut(&Message) + Send + 'static;
-type OnError = dyn FnMut(&Error) + Send + 'static;
-type OnDone = dyn FnMut(Option<&Error>) + Send + 'static;
-
-unsafe fn pchar_to_string(p_value: *const c_char) -> String {
-    assert!(!p_value.is_null());
-    let value = String::from(CStr::from_ptr(p_value).to_str().unwrap());
-    value
-}
-
-enum OwnedOrBorrowedError {
-    Owned(*mut mdb_sys::mdb_error_t),
-    Borrowed(*const mdb_sys::mdb_error_t),
-}
-pub struct Error(OwnedOrBorrowedError);
+pub trait Callback<T>: FnMut(T) + Send + 'static {}
+impl<T, F> Callback<T> for F where F: FnMut(T) + Send + 'static {}
 
-impl Debug for Error {
-    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
-        write!(f, "Code: {}; Message: {:?};", self.code(), self.message())
-    }
-}
-
-impl Display for Error {
-    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
-        write!(f, "{} ({})", self.code(), self.message())
-    }
-}
-
-impl Drop for Error {
-    fn drop(&mut self) {
-        unsafe {
-            match &mut self.0 {
-                OwnedOrBorrowedError::Owned(ptr) => {
-                    mdb_sys::mdb_error_destroy(ptr);
-                }
-                OwnedOrBorrowedError::Borrowed(_) => {}
-            }
-        }
-    }
-}
-unsafe impl Send for Error {}
-// Note that error is not Sync
-
-impl std::error::Error for Error {}
-
-impl Error {
-    fn new_owned(ptr: *mut mdb_sys::mdb_error_t) -> Self {
-        assert!(!ptr.is_null());
-        Self(OwnedOrBorrowedError::Owned(ptr))
-    }
-    fn new_borrowed(ptr: *const mdb_sys::mdb_error_t) -> Self {
-        assert!(!ptr.is_null());
-        Self(OwnedOrBorrowedError::Borrowed(ptr))
-    }
-    fn as_ref(&self) -> &mdb_sys::mdb_error_t {
-        unsafe {
-            match self.0 {
-                OwnedOrBorrowedError::Owned(ptr) => ptr.as_ref(),
-                OwnedOrBorrowedError::Borrowed(ptr) => ptr.as_ref(),
-            }
-        }
-        .unwrap()
-    }
-
-    fn code(&self) -> i32 {
-        self.as_ref().code
-    }
-
-    fn message(&self) -> String {
-        unsafe { pchar_to_string(self.as_ref().message) }
-    }
-}
+type OnMessage = dyn for<'a> Callback<Message<'a>>;
+type OnError = dyn for<'a> Callback<BorrowedError<'a>>;
+type OnDone = dyn for<'a> Callback<Option<BorrowedError<'a>>>;
 
 pub struct Connection {
     ptr: *mut mdb_sys::mdb_connection_t,
-    on_error: *mut Box<OnError>,
+    on_error: *mut OnError,
 }
 
 impl Connection {
     // TODO: Consider adopting a builder-like pattern.
-    pub fn try_new(on_error: Option<Box<OnError>>) -> Result<Self, Error> {
+    /// Passing `None` as the `@on_error` callback requires qualifying the generic like so:
```

Comment:
As far as I know it is convention in Rust to reference parameters without an `@` prefix, see for instance `Box::new`.